<a href="https://colab.research.google.com/github/89rael/ml-examples/blob/main/Classifier_with_Scikit_learn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import sklearn

In [13]:
import pandas as pd

In [23]:
from sklearn.model_selection import train_test_split

In [27]:
from sklearn.naive_bayes import GaussianNB

In [39]:
from sklearn.metrics import accuracy_score

The dataset we will be working with is the [Breast Cancer Wisconsin Diagnostic Database](https://scikit-learn.org/stable/datasets/index.html#breast-cancer-wisconsin-diagnostic-database). The dataset includes various information about breast cancer tumors, as well as classification labels of malignant or benign. The dataset has 569 instances and includes information on 30 attributes, or features, such as the radius of the tumor, texture smoothness, and area.

In [3]:
# import and load the dataset included in scikit-learn
from sklearn.datasets import load_breast_cancer

In [4]:
data = load_breast_cancer()

In [22]:
type(data)

sklearn.utils.Bunch

## What’s in our Dictionary (Bunch)?

Scikit’s dictionary or Bunchis really powerful. Let’s begin this dictionary by looking at its keys.

`data.keys()`

We get the following keys:

*  *`data`* is all the feature data (the attributes of the scan that help us identify if the tumor is malignant or benign, such as radius, area, etc.) in a NumPy array
*  *`target`* is the target data (the variable you want to predict, in this case whether the tumor is malignant or benign) in a NumPy array,

These two keys are the actual data. The remaining keys (below), serve a descriptive purpose. It’s important to note that all of Scikit-Learn datasets are divided into data and target. data represents the features, which are the variables that help the model learn how to predict. target includes the actual labels. In our case, the target data is one column classifies the tumor as either 0 indicating malignant or 1 for benign.

*  *`feature_names`* are the names of the feature variables, in other words names of the columns in data
*  *`target_names`* is the name(s) of the target variable(s), in other words name(s) of the target column(s)
* *`DESCR`* , short for DESCRIPTION, is a description of the dataset
*  *`filename`* is the path to the actual file of the data in CSV format.

In [17]:
# create a pandas df with the features
df = pd.DataFrame(data.data, columns=data.feature_names)

In [19]:
# adding the target column
df['target'] = data.target

In [20]:
df.head()

mean radius  mean texture  ...  worst fractal dimension  target
0        17.99         10.38  ...                  0.11890       0
1        20.57         17.77  ...                  0.08902       0
2        19.69         21.25  ...                  0.08758       0
3        11.42         20.38  ...                  0.17300       0
4        20.29         14.34  ...                  0.07678       0

[5 rows x 31 columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

#### Now we heve to split our data into training set and test set.

In [26]:
train, test = train_test_split(df,test_size=0.33, random_state=42, shuffle=True)

## Building and Evaluating the Model

There are many models for machine learning, and each model has its
own strengths and weaknesses. In this case, we will focus on a simple
algorithm that usually performs well in binary classification tasks,
namely [Naive Bayes (NB)](https://scikit-learn.org/stable/modules/naive_bayes.html).

In [28]:
# initialize our classifier
gnb = GaussianNB()

In [35]:
# train our classifier
model = gnb.fit(train.loc[:, :'worst fractal dimension'], train['target'])

Now we can use the trained model to make predictions on our test set, which we do using the `predict()` function.
The `predict()` function returns an array of predictions for each data
instance in the test set. 

In [37]:
# make predictions
preds = gnb.predict(test.loc[:, :'worst fractal dimension'])
preds

array([1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1])

##  Evaluating the Model’s Accuracy

Using the target column in test dataframe, we can evaluate the accuracy of our
model’s predicted values by comparing the two series `(target column
vs. preds)`. We will use the sklearn function `accuracy_score()` to
determine the accuracy of our machine learning

In [40]:
# evaluate accuracy
accuracy_score(test['target'], preds)

0.9414893617021277

The NB classifier is 94.15% accurate. This means that 94.15 percent of the time the classifier is able to make the correct prediction as to whether or not the tumor is malignant or benign.These results suggest that our feature set of 30 attributes are good indicators of tumor class